### PR TITLE
feat(connector): [Globepay] add authorize and psync flow 

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added support for Globepay payment gateway
Payment Method: WechatPay and Alipay (Both QrCodeFlow)
New payment Method : Alipay(**Note : Qr code url is given in the next action, this qr code has to be shown in sdk.**)
Supported payment flows

Authorize
PSync
**Note** : It doesn't support manual capture

Payment method data
Screenshot 2023-07-13 at 2 31 06 PM
Screenshot 2023-07-13 at 2 30 51 PM

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?

local testing as it is not a redirection flow , we can't write ui_tests
Wechatpay
Screenshot 2023-07-06 at 6 51 56 PM
Alipay
Screenshot 2023-07-06 at 6 52 22 PM
Qrcode that need be displayed in sdk
Screenshot 2023-07-06 at 6 54 39 PM
ightly fmt --all`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/globepay.rs [hunk 1]
```diff
@@ -55,21 +56,52 @@ where
...
-        let mut api_key = self.get_auth_header(&req.connector_auth_type)?;
-        header.append(&mut api_key);
         Ok(header)
     }
 }
 
+fn get_globlepay_query_params(
+    connector_auth_type: &types::ConnectorAuthType,
+) -> CustomResult<String, errors::ConnectorError> {
+    let api_key = get_globepay_api_key(connector_auth_type)?;
```
Comment (prasunna09) [inline crates/router/src/connector/globepay.rs:73]: Rename it to auth_type as you are returning "connector_auth_type"

File: crates/router/src/connector/globepay.rs [hunk 2]
```diff
@@ -85,14 +117,9 @@ impl ConnectorCommon for Globepay {
...
-        auth_type: &types::ConnectorAuthType,
+        _auth_type: &types::ConnectorAuthType,
     ) -> CustomResult<Vec<(String, request::Maskable<String>)>, errors::ConnectorError> {
-        let auth = globepay::GlobepayAuthType::try_from(auth_type)
-            .change_context(errors::ConnectorError::FailedToObtainAuthType)?;
-        Ok(vec![(
-            headers::AUTHORIZATION.to_string(),
-            auth.api_key.expose().into_masked(),
-        )])
+        Ok(vec![])
```
Comment (prasunna09) [inline crates/router/src/connector/globepay.rs:122]: Please remove this function as you are not using it anywhere in the code

File: crates/router/src/connector/globepay/transformers.rs [hunk 3]
```diff
@@ -1,90 +1,118 @@
...
-                })
-            }
-            _ => Err(errors::ConnectorError::NotImplemented("Payment methods".to_string()).into()),
-        }
+        let channel: GlobepayChannel = match &item.request.payment_method_data {
+            api::PaymentMethodData::Wallet(ref wallet_data) => match wallet_data {
+                api::WalletData::AliPay(_) => GlobepayChannel::Alipay,
+                api::WalletData::WeChatPay(_) => GlobepayChannel::Wechat,
+                _ => Err(errors::ConnectorError::NotImplemented(
+                    "Payment methods".to_string(),
```
Comment (prasunna09) [inline crates/router/src/connector/globepay/transformers.rs:32]: `"Payment method".to_string(),`

Change it in other places too!

**Review by prasunna09 — CHANGES_REQUESTED** [review]

**Review by prasunna09 — APPROVED** [review]

File: crates/router/src/connector/globepay/transformers.rs [hunk 4]
```diff
@@ -1,90 +1,118 @@
...
+        let description =
+            item.description
+                .clone()
+                .ok_or(errors::ConnectorError::MissingRequiredField {
+                    field_name: "description",
+                })?;
+        Ok(Self {
+            price: item.request.amount,
+            description,
+            currency: item.request.currency,
```
[the remark this answers is not in the document]
Comment (prasunna09) [reply crates/router/src/connector/globepay/transformers.rs:48]: Please do not add this currency in required field, as we had a talk with sdk team. Please remove if added

File: crates/router/src/connector/globepay/transformers.rs [hunk 5]
```diff
@@ -100,21 +128,127 @@ impl<F, T>
...
+            Ok(Self {
+                status: enums::AttemptStatus::Failure, //As this connector gives 200 in failed scenarios . if return_code is not success status is mapped to failure. ref = "https://pay.globepay.co/docs/en/#api-QRCode-NewQRCode"
+                response: Err(types::ErrorResponse {
+                    code: item.response.return_code.to_string(),
+                    message: item.response.return_code.to_string(),
+                    reason: item.response.return_msg,
+                    status_code: item.http_code,
+                }),
+                ..item.data
+            })
```
Comment (jagan-jaya) [inline crates/router/src/connector/globepay/transformers.rs:173]: move to a method and call it here

File: crates/router/src/connector/globepay/transformers.rs [hunk 6]
```diff
@@ -100,21 +128,127 @@ impl<F, T>
...
+                status: enums::AttemptStatus::Failure,
+                response: Err(types::ErrorResponse {
+                    code: item.response.return_code.to_string(),
+                    message: item.response.return_code.to_string(),
+                    reason: item.response.return_msg,
+                    status_code: item.http_code,
+                }),
+                ..item.data
+            })
+        }
```
Comment (jagan-jaya) [inline crates/router/src/connector/globepay/transformers.rs:247]: reuse it by moving to a new method

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/globepay/transformers.rs [hunk 7]
```diff
@@ -100,21 +124,138 @@ impl<F, T>
...
+                ..item.data
+            })
+        } else {
+            Ok(Self {
+                status: enums::AttemptStatus::Failure, //As this connector gives 200 in failed scenarios . if return_code is not success status is mapped to failure. ref = "https://pay.globepay.co/docs/en/#api-QRCode-NewQRCode"
+                response: Err(get_error_response(
+                    item.response.return_code,
+                    item.response.return_msg,
+                    item.http_code,
+                )),
```
Comment (jagan-jaya) [inline crates/router/src/connector/globepay/transformers.rs:166]:
```suggestion
                response: get_error_response(item),
```

File: crates/router/src/connector/globepay/transformers.rs [hunk 8]
```diff
@@ -100,21 +124,138 @@ impl<F, T>
...
```
Comment (jagan-jaya) [inline crates/router/src/connector/globepay/transformers.rs:238]:
```suggestion
                response: get_error_response(item),
```

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/globepay/transformers.rs [hunk 9]
```diff
@@ -1,90 +1,118 @@
...
+    Exists,
 }
 
 impl From<GlobepayPaymentStatus> for enums::AttemptStatus {
     fn from(item: GlobepayPaymentStatus) -> Self {
         match item {
-            GlobepayPaymentStatus::Succeeded => Self::Charged,
-            GlobepayPaymentStatus::Failed => Self::Failure,
-            GlobepayPaymentStatus::Processing => Self::Authorizing,
+            GlobepayPaymentStatus::Success => Self::AuthenticationPending,
```
[the remark this answers is not in the document]
Comment (srujanchikke) [reply crates/router/src/connector/globepay/transformers.rs]: added comments!

File: crates/router/src/connector/globepay/transformers.rs (hunk 7 above)
Comment (srujanchikke) [reply crates/router/src/connector/globepay/transformers.rs:166]: we can't pass `item` as `item.response` is different in different flows (PaymentResponse, PaymentSyncResponse)

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/globepay/transformers.rs [hunk 10]
```diff
@@ -100,21 +124,138 @@ impl<F, T>
...
+}
+
+fn get_error_response(
+    return_code: GlobepayReturnCode,
+    return_msg: Option<String>,
+    status_code: u16,
+) -> types::ErrorResponse {
+    types::ErrorResponse {
+        code: return_code.to_string(),
+        message: return_code.to_string(),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/globepay/transformers.rs:252]: Let's set No_error_message for message

File: crates/router/src/connector/globepay.rs [hunk 11]
```diff
@@ -106,17 +118,16 @@ impl ConnectorCommon for Globepay {
 
         Ok(ErrorResponse {
             status_code: res.status_code,
-            code: response.code,
-            message: response.message,
-            reason: response.reason,
+            code: response.return_code.to_string(),
+            message: response.return_code.to_string(),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/globepay.rs:122]: Let's set No_error_message for message

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**